### DISCUSSION TITLE
How to optimize reloadLookup() and .getLookup() ?

### DISCUSSION THREAD

**Discussion by ibrahimGit123** [serenity-is/Serenity#6679] [Help]
I have found lookups to be much slower since upgrading Serenity. How do I optimize these lookups lines of code in grids and dialogs? 
`Q.reloadLookup(table.lookupKey);`
`Row.getLookup().itemById[Id];`

These run extremely slow on any table with a few thousand rows (I know this is a known issue, but it is even slower than before now).

I have already managed to to optimize lookup editors in row.cs by using the attribute ServiceLookupEditor instead of LookupEditor.

**Comment by edwardch** [comment]
If you have lots of records, the ServiceLookupEditor will help. Otherwise you can reduce the number of columns in your lookupinclude, and you could even reduce the names of the columns to reduce the overall json response.

**Comment by LSNicholls** [comment]
Do either of you know what the recommended replacement for getLookup() is, considering that it is marked deprecated?  

UserGrid has this line of code:

      ``  var roleList = (ctx.value || []).map(x => (RoleRow.getLookup().itemById[x] || {}).RoleName || "");  
      
... and many of the samples that show the results of multiple selection in a grid do the same thing. But the build warnings indicate that this is no longer optimal:

`Warning	TS6387	(TS) The signature '(): Lookup<RoleRow>' of 'RoleRow.getLookup' is deprecated.`

I tried just replacing it like this:

`var roleList = (ctx.value || []).map(async x => ((await (RoleRow.getLookupAsync())).itemById[x]  || {}).RoleName || "");  `

... but that results in [Object Promise] in the grid rather than the delimited string of RoleNames.

Am I on the right track?  Or is there a different solution?

Obviously these types of lookups are not being done where there are thousands of records, so it isn't really a perf problem.  I just don't think that the signature would be deprecated unless there is a replacement strategy to use, so I think there must be one!

Thanks in advance for your thoughts.

**Reply by furkanevran** [reply]
@LSNicholls Hi, [this document](https://serenity.is/docs/framework/dynamic-scripts/lookups) explains why `getLookup()` was deprecated.

You could refer to the diff below about how to refactor UserGrid to use `getLookupAsync()`.

```diff
--- a/src/App.Web/Modules/Administration/User/UserRow.cs
+++ b/src/App.Web/Modules/Administration/User/UserRow.cs
     [DisplayName("Roles"), LinkingSetRelation(typeof(UserRoleRow), nameof(UserRoleRow.UserId), nameof(UserRoleRow.RoleId))]
-    [LookupEditor(typeof(RoleRow), Multiple = true)]
+    [ServiceLookupEditor(typeof(RoleRow), Multiple = true, Async = true)]
     public List<int> Roles
     {
         get => fields.Roles[this];
```

```diff
--- a/src/App.Web/Modules/Administration/User/UserGrid.ts
+++ b/src/App.Web/Modules/Administration/User/UserGrid.ts
@@ -1,5 +1,5 @@
 import { Decorators, EntityGrid } from "@serenity-is/corelib";
-import { resolveUrl, tryFirst } from "@serenity-is/corelib/q";
+import { getLookup, resolveUrl, tryFirst } from "@serenity-is/corelib/q";
 import { RoleRow, UserColumns, UserRow, UserService } from "../";
 import { UserDialog } from "./UserDialog";
 
@@ -36,8 +36,17 @@
 export class UserGrid extends EntityGrid<UserRow, any> {
 
         var roles = tryFirst(columns, x => x.field == UserRow.Fields.Roles);
         if (roles) {
+            var rolesLookup: Q.Lookup<RoleRow>;
+            RoleRow.getLookupAsync().then(lookup => {
+                rolesLookup = lookup;
+                this.slickGrid.invalidate();
+            });
+
             roles.format = ctx => {
-                var roleList = (ctx.value || []).map(x => (RoleRow.getLookup().itemById[x] || {}).RoleName || "");
+                if (!rolesLookup)
+                    return `<i class="fa fa-spinner"></i>`;
+
+                var roleList = (ctx.value || []).map(x => (rolesLookup.itemById[x] || {}).RoleName || "");
                 roleList.sort();
                 return roleList.join(", ");
             };
```

**Reply by LSNicholls** [reply]
Hi @furkanevran 

I didn't really have any problem understanding **why** it was deprecated, just didn't know how to replace it :-).  Thank you!  And it is very helpful to realize that I could go to the source and diff, assuming the problem was corrected later than my copy of the file was extracted, and find this out for myself in the future.
